crates/solver/src/driver/solver_settlements.rs

```diff
@@ -90,30 +93,148 @@ fn merge_at_most_settlements(
     }
 }
 
-pub fn filter_settlements_without_old_orders(
+/// Filters out all settlements without any orders which are mature by age or mature by association.
+/// Any order older than `min_order_age` is considered to be mature by age.
+/// Any younger order in a settlement containing an order mature by age or mature by association
+/// is considered to be mature by association.
+/// Because maturity by association is defined recursively it can "spread" across settlements,
+/// resulting in settlements being allowed where it's not immediately obvious by which association
+/// any order of a settlement has matured.
+pub fn retain_mature_settlements(
     min_order_age: Duration,
-    settlements: &mut Vec<Settlement>,
-) {
-    let settle_orders_older_than =
-        chrono::offset::Utc::now() - chrono::Duration::from_std(min_order_age).unwrap();
-    settlements.retain(|settlement| {
-        settlement
-            .trades()
-            .iter()
-            .any(|trade| trade.order.order_meta_data.creation_date <= settle_orders_older_than)
-    });
+    settlements: Vec<Settlement>,
+) -> Vec<Settlement> {
+    fn find_mature_settlements(
+        min_order_age: Duration,
+        settlements: &[Settlement],
+    ) -> HashSet<usize> {
+        let settle_orders_older_than =
+            chrono::offset::Utc::now() - chrono::Duration::from_std(min_order_age).unwrap();
+
+        let mut valid_trades = HashSet::<&model::order::OrderUid>::default();
+        let mut valid_settlement_indices = HashSet::<usize>::default();
+
+        loop {
```

Comment:
nanonit: might have been nice to upper bound this loop (intuitively I'd say by `settlements.length`) to make it more obvious that this can not loop endlessly (I'm always scared when I read unbounded loops).

Reply:
That's a good point. :+1:
It didn't occur to me that giving an upper bound (while not changing the correctness of the solution) would help convey more information about the runtime and give peace of mind.
I will keep an eye out for `loop{}` in the future and see whether I could add more information that way.